File: problems/01_matrix/solution.py
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        if not matrix: return matrix
        M, N = len(matrix), len(matrix[0])
        seen = set()
        q = []
        
        # get all zeros O(matrix)
        for y in range(M):
            for x in range(N):
                if matrix[y][x] == 0:
                    seen.add((y, x))
                    q.append((y, x))
                    
        dir_ = ((1, 0), (-1, 0), (0, 1), (0, -1))
        
        # flood-fill all non-zero cells O(matrix)
        cnt = 0
        while q:
            nq = []
            cnt += 1
            for y, x in q:
                for ny, nx in ((y+r, x+c) for r, c in dir_):
                    if 0 <= ny < M and 0 <= nx < N and (ny, nx) not in seen:
                        seen.add((ny, nx))
                        matrix[ny][nx] = cnt
                        nq.append((ny, nx))
            q = nq        
        return matrix

File: problems/01_matrix/solution.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int R = mat.size(), C = mat[0].size();
        vector<vector<bool>> seen(R, vector<bool>(C, false));
        queue<pair<int, int>> q;
        for (int i = 0; i < R; ++i) {
            for (int j = 0; j < C; ++j) {
                if (mat[i][j] == 0) {
                    q.push({i, j});
                    seen[i][j] = true;
                }
            }
        }
        int c = 0;
        vector<int> dir{-1, 0, 1, 0, -1};
        while (!q.empty()) {
            int len = q.size();
            for (int i = 0; i < len; ++i) {
                auto [y, x] = q.front();
                mat[y][x] = c;
                q.pop();
                for (int j = 0; j < 4; ++j) {
                    int ny = y + dir[j], nx = x + dir[j+1];
                    if (0 <= ny && ny < R && 0 <= nx && nx < C) {
                        if (!seen[ny][nx]) {
                            seen[ny][nx] = true;
                            q.push({ny, nx});
                        }
                    }
                }
            }
            ++c;
        }
        return mat;
    }
};